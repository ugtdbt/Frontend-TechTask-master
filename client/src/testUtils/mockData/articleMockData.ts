export const article = {
  imageUrl:
    "https://cdn1.home24.net/images/media/catalog/product/200x200/png/l/a/lattenrost-super-fix-90-x-200cm-4448116.webp",
  name: "Lattenrost Super Fix",
  variantName: "Breite: 140 cm",
  price: 14990,
};
