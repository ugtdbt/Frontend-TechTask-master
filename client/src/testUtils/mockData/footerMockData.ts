export const footer = {
  description:
    "Alle Preise sind in Euro (â‚¬) inkl. gesetzlicher Umsatzsteuer und Versandkosten.",
};
