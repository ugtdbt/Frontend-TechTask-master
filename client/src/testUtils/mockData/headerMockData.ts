export const sections = [
  { title: "furniture", url: "#" },
  { title: "lamps", url: "#" },
  { title: "accessories", url: "#" },
  { title: "textiles", url: "#" },
  { title: "children", url: "#" },
  { title: "kitchen", url: "#" },
  { title: "bath", url: "#" },
  { title: "Garden", url: "#" },
  { title: "Brands", url: "#" },
  { title: "Shop the look", url: "#" },
  { title: "advisory", url: "#" },
  { title: "stores", url: "#" },
  { title: "Sale", url: "#" },
];

export const header = {
  sections,
  title: "Home24",
};
