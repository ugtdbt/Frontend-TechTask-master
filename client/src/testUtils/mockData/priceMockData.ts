export const price = {
  price: 2500,
};
