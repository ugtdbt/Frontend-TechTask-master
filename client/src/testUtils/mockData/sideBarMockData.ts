export const sideBar = {
  list: [
    { name: "Wohnzimmer", urlPath: "wohnzimmer-moebel/" },
    { name: "Polsterm√∂bel", urlPath: "alle-polstermoebel/" },
  ],
};
