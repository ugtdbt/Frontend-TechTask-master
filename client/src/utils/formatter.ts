const intlNumberFormatValues = ["de-DE", "currency", "EUR"];

/**
 * @description Format currency base on de-DE
 */

export const formatter = new Intl.NumberFormat(intlNumberFormatValues[0], {
  style: intlNumberFormatValues[1],
  currency: intlNumberFormatValues[2],
});
