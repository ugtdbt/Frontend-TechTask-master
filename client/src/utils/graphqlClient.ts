import { GraphQLClient } from "graphql-request";

/**
 * @description Format currency base on de-DE
 */

export const graphqlClient = new GraphQLClient("/graphql", {});
