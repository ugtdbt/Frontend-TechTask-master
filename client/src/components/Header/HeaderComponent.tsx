import React, { memo, useState } from "react";
import type { FC } from "react";
import Toolbar from "@mui/material/Toolbar";
import SearchIcon from "@mui/icons-material/Search";
import Typography from "@mui/material/Typography";
import Link from "@mui/material/Link";
import Grid from "@mui/material/Grid";
import IconButton from "@mui/material/IconButton";
import ShoppingCartIcon from "@mui/icons-material/ShoppingCart";
import FavoriteIcon from "@mui/icons-material/Favorite";
import Autocomplete from "@mui/material/Autocomplete";
import { useInfiniteSearchSuggestionsV2Query } from "../../gql/product.generated-query";
import { graphqlClient } from "../../utils/graphqlClient";

import {
  Search,
  SearchIconWrapper,
  StyledTextField,
} from "../../styles/header";
import { ArticleImageFormat, Locale, SearchBackend } from "../../gql";

interface HeaderProps {
  sections: ReadonlyArray<{
    title: string;
    url: string;
  }>;
  title: string;
}

// ToDo - Write handle navigation to homepage

const Header: FC<HeaderProps> = memo((props: HeaderProps) => {
  const { sections, title } = props;
  const [searchValues, setSearchValues] = useState<string[]>([]);
  const [query, setQuery] = useState<string>("");

  const handleSearch = (searchText: string): void => {
    if (searchText === null) return;
    setQuery(searchText);
    refetch().catch(() => {
      setSearchValues([]);
    });
  };

  const { refetch } = useInfiniteSearchSuggestionsV2Query(
    "prefix",
    graphqlClient,
    {
      prefix: query,
      locale: Locale.DeDe,
      backend: SearchBackend.ThirdParty,
      format: ArticleImageFormat.Webp,
    },
    {
      onSuccess: (data) => {
        let searchSuggestions: any[] = [];
        if (typeof data?.pages[0]?.suggestions !== "undefined") {
          if (data?.pages[0]?.suggestions !== null) {
            searchSuggestions = data?.pages[0]?.suggestions
              .filter((suggestion) => {
                return typeof suggestion?.name !== "undefined";
              })
              .map((value) =>
                typeof value?.name !== "undefined" ? value.name : ""
              );
          }
        }
        setSearchValues(searchSuggestions);
      },
      onError: () => {
        setSearchValues([]);
      },
    }
  );

  return (
    <React.Fragment>
      <Toolbar sx={{ borderBottom: 1, borderColor: "divider" }}>
        <Grid container>
          <Grid item xs={12} md={2}>
            <Typography
              component="h2"
              variant="h5"
              color="inherit"
              align="left"
              noWrap
              sx={{ flex: 1 }}
              data-testid="header-title"
            >
              {title}
            </Typography>
          </Grid>
          <Grid item xs={12} md={8}>
            <Search>
              <SearchIconWrapper>
                <SearchIcon />
              </SearchIconWrapper>
              <Autocomplete
                freeSolo
                disableClearable
                options={[...searchValues]}
                onInputChange={(event, newInputValue) => {
                  handleSearch(newInputValue);
                }}
                renderInput={(params) => (
                  <StyledTextField
                    {...params}
                    InputProps={{
                      ...params.InputProps,
                      type: "search",
                    }}
                  />
                )}
              />
            </Search>
          </Grid>
          <Grid item xs={12} md={2} pl={3} justifyContent="flex-end">
            <Grid container justifyContent="flex-end">
              <IconButton aria-label="add to favorites">
                <FavoriteIcon />
              </IconButton>
              <IconButton aria-label="share">
                <ShoppingCartIcon />
              </IconButton>
            </Grid>
          </Grid>
        </Grid>
      </Toolbar>
      <Toolbar
        component="nav"
        variant="dense"
        sx={{ justifyContent: "space-between", overflowX: "auto" }}
      >
        {sections.map((section) => (
          <Link
            color="inherit"
            noWrap
            key={section.title}
            variant="body2"
            href={section.url}
            sx={{ p: 1, flexShrink: 0 }}
          >
            {section.title}
          </Link>
        ))}
      </Toolbar>
    </React.Fragment>
  );
});

Header.displayName = "Header";
export default Header;
