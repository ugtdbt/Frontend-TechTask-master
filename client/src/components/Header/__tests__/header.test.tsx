import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { header } from "../../../testUtils/mockData/headerMockData";
import Header from "../HeaderComponent";
import { queryClient } from "../../../utils/react-query-client";
import { QueryClientProvider } from "@tanstack/react-query";

describe("Header component", () => {
  const HeaderComponent = (
    <QueryClientProvider client={queryClient}>
      <Header sections={header.sections} title={header.title} />
    </QueryClientProvider>
  );

  afterEach(() => {
    cleanup();
  });

  it("Header Match Snapshot", () => {
    const tree = renderer.create(HeaderComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Check Header title", () => {
    render(HeaderComponent);
    const headerTitle = screen.getByTestId("header-title")?.innerHTML;
    expect(headerTitle).toContain(header.title);
  });
});
