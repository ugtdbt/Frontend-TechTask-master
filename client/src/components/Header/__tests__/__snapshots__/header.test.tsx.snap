// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Header Match Snapshot 1`] = `
Array [
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-16kjkpc-MuiToolbar-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-2 css-12xeub0-MuiGrid-root"
      >
        <h2
          className="MuiTypography-root MuiTypography-h5 MuiTypography-alignLeft MuiTypography-noWrap css-1m4x84m-MuiTypography-root"
          data-testid="header-title"
        >
          Home24
        </h2>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-9drkqd-MuiGrid-root"
      >
        <div
          className="css-1wrvpbt"
        >
          <div
            className="css-1iz4iln"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <div
            aria-owns={null}
            className="MuiAutocomplete-root css-1qqsdnr-MuiAutocomplete-root"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
          >
            <div
              className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1y2of3x-MuiFormControl-root-MuiTextField-root"
            >
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiAutocomplete-inputRoot css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                onClick={[Function]}
              >
                <input
                  aria-activedescendant={null}
                  aria-autocomplete="list"
                  aria-expanded={false}
                  aria-invalid={false}
                  autoCapitalize="none"
                  autoComplete="off"
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch MuiAutocomplete-input MuiAutocomplete-inputFocused css-1g24dm6-MuiInputBase-input-MuiOutlinedInput-input"
                  disabled={false}
                  id=":r0:"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onMouseDown={[Function]}
                  readOnly={false}
                  required={false}
                  role="combobox"
                  spellCheck="false"
                  type="search"
                  value=""
                />
                <fieldset
                  aria-hidden={true}
                  className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="css-ihdtdm"
                  >
                    <span
                      className="notranslate"
                    >
                      â€‹
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-2 css-1y1l3sl-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
        >
          <button
            aria-label="add to favorites"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FavoriteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
              />
            </svg>
          </button>
          <button
            aria-label="share"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ShoppingCartIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <nav
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-dense css-14105se-MuiToolbar-root"
  >
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      furniture
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      lamps
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      accessories
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      textiles
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      children
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      kitchen
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      bath
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      Garden
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      Brands
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      Shop the look
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      advisory
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      stores
    </a>
    <a
      className="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiLink-root MuiLink-underlineAlways css-1a72jx6-MuiTypography-root-MuiLink-root"
      href="#"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      Sale
    </a>
  </nav>,
]
`;
